Serializer updates

rshar1 opened the pull request:
# The Serializer:

- Overloaded the write and read methods to include a new write and read that instead will use a printwriter to write and a buffered reader to read. 
- These methods were overloaded, rather than being completely replaced, so that communication with the relay can continue using the input and output streams.
- All the serializers have also been modified to correctly implement the new serializer interface

## Client Side:
Controller and view have been modified to use the new write and read methods of the serializers

## Server Side:
Server has been updated to use the new write and read methods, as well.

rshar1 committed:
Added new methods to serializer

Added two new methods to Serializer Interface. These methods will allow
messages to be send using a print writer, rather than an byte streams.

rshar1 committed:
implemented the read and write of serializers

Implemented the read and write methods of the serializers class.

rshar1 committed:
removed all todos

Tested code in Serializers and removed old todos

rshar1 committed:
implemented Uuids serializers

Implemented the serializers in the Uuids. Testing still needed,
however.

rshar1 committed:
modified relay to allow compilation

Added write and read methods to relay code to allow compilation after
modification of the Serializer interface

rshar1 committed:
updated the serializer

Updated the Serializer interface to use a buffered reader instead of a
StringBuffer with delimiter. The delimiter will be the new line.

rshar1 committed:
updated uuids serializer

Updated the serializer in uuids to use the printwriter and buffered
reader.

rshar1 committed:
updated all serializers

Updated all the remaining serializers to use the BufferedReader and
PrintWriter, rather than the input and output streams

rshar1 committed:
updated Relay for compilation

Updated the serializers in the relay so the code can compile correctly.

rshar1 committed:
Server now uses new Serializer

The server now uses the bufferedReader and print writer to communicate
with the client.

rshar1 committed:
fixed auto flush on printwriter

Added auto flush to the print writer.

rshar1 committed:
Merge remote-tracking branch 'origin/testing' into serializerUpdates

# Conflicts:
#	src/codeu/chat/client/Controller.java
#	src/codeu/chat/client/View.java
#	src/codeu/chat/server/Server.java
#	src/codeu/chat/util/Uuid.java

rshar1 committed:
added back the travis yml

Added back the travis.yml which was accident removed when pulling the
changes from testing branch.

turtlefarmer commented on src/codeu/chat/client/Controller.java:
```diff
@@ -46,13 +48,17 @@ public Message newMessage(Uuid author, Uuid conversation, String body) {
-      Serializers.STRING.write(connection.out(), body);
+      PrintWriter out = new PrintWriter(connection.out(), true);
+      BufferedReader in = new BufferedReader(new InputStreamReader(connection.in()));
+      
```
Overall good change so far to use println and readln so you don't have to count bytes.  An extension of this work that makes sense to me is to move towards line-break as a message delimiter instead of a message-variable delimiter.  As in, find a way to pack an entire message (whether it's new_message.body+author+conversation or something else) into a single line.  That way, if you add a uuid of some sort fo yoru request/response pairs, the request -> response mechanism can be asynchronous.

turtlefarmer replied:
One easy way to package objects into single line is through some sort of json serializer.  I'm sure there's a library out there that does this for you somewhere that just serializes all public variables.

turtlefarmer replied:
(just to be clear, this isn't blocking for this pull request, just an idea)

turtlefarmer commented on src/codeu/chat/util/Uuid.java:
```diff
@@ -68,6 +68,44 @@ public Uuid read(InputStream in) throws IOException {
+      Uuid head = null;
+
+      for (int i = length - 1; i >= 0; i--) {
+        head = new Uuid(head, chain[i]);
```
is this right?

rshar1 replied:
I believe so...I used the provided code as a template.

rshar1 committed:
Merge branch 'testing' into serializerUpdates

rshar1 committed:
added gson library

rshar1 committed:
corrected shell commands

Corrected the shell commands so the application will now successfully
compile and execute with the gson library

rshar1 committed:
updated all the serializers to use json

Updated the serializers for Conversation, conversationSummary, Message,
User, and Uuid to use the gson library.

rshar1 committed:
removed unnecessary inputs

rshar1 committed:
Merge branch 'jsonSerializer' into serializerUpdates

rshar1 committed:
updated test.sh

updated test.sh to include the gson library in the class path

rshar1 committed:
added uuid tests

Added tests to test the json serializer works properly for the uuid

rshar1 committed:
corrected Uuid Test

Corrected Uuid Test to use the Serializers instead of directly writing
to the stream.

rshar1 committed:
created time json serializer

Implemented a custom serializer for the time object so ms are no lost.

Added the Gson object to the Serializers class so it can be accessed by
all the serializers.

rshar1 committed:
Cleaned code and completed json serializer

rshar1 committed:
Added tests for Time and Conversation

Added tests to test the serialization of conversation and Time.

turtlefarmer commented:
looks good to merge :)

rshar1 committed:
Added test for message serializer

Added a test to test the json serialization of message
